### ISSUE TITLE
Why my lint-staged output is different when is triggered by Husky?

### ISSUE THREAD

**Issue by luquera** [typicode/husky#968]
Everything is set up correctly, according to the documentation. I'm using Windows 10.

by husky:
![image](https://user-images.githubusercontent.com/20865763/118204066-c5374e80-b433-11eb-82bc-9c44239230ca.png)

by npm script (running lint-staged):
![image](https://user-images.githubusercontent.com/20865763/118204260-3119b700-b434-11eb-9b0c-fb5678b9f49f.png)

Thanks!

**Comment by walkerrandolphsmith** [comment]
I am experiencing this as well with lint staged 11.0.0 and husky v6. In addition, when I run `yarn lint-staged` it updates staged files, but when committing and relying on a hook to call `yarn lint-staged` the changes are left un-staged and a commit is made.

**Comment by MineiToshio** [comment]
Having the same issue, @luquera did you manage to solve it?

**Comment by polkovnikov-ph** [comment]
Well, unstale.

**Comment by luquera** [comment]
> Having the same issue, @luquera did you manage to solve it?

Unfortunately not yet :(

**Comment by FFdhorkin** [comment]
Issue is still not solved

**Comment by mukhlisakbr** [comment]
it still not solved till now

**Comment by polkovnikov-ph** [comment]
bump

**Comment by typicode** [comment]
Not sure, but I guess it detects that the terminal aren't capable of rendering the first version and renders the second one. I imagine it's done via an env variable. I would try passing some env variable to try if it's possible to "force" the first type of rendering.

And sometimes, it's better to avoid animation and have per-line information (for example logs in Git GUIs).

I would check the dependencies that lint-staged uses for rendering the results.

That said, as far as I know, I don't think there's anything that can be done in husky's code to help with that.

**Comment by TrejGun** [comment]
same here

**Comment by acidcrash2005** [comment]
Has anyone found resolve?

**Comment by AndysonDK** [comment]
It worked for me before but since yesterday I've now bumped into the same issue, I'm running WSL2 using the windows terminal and using huske to run lint-staged

**Comment by valentinoli** [comment]
https://github.com/okonet/lint-staged/issues/693#issuecomment-1079759224

**Comment by iiroj** [comment]
The issue seems to be that `supports-color` ([code link](https://github.com/chalk/supports-color/blob/main/index.js#L165)) and `colorette` ([code link](https://github.com/jorgebucaran/colorette/blob/main/index.js#L11)) both use the `tty.isatty(1)` check from `node:tty` to test whether the current stdin is in fact a TTY. This is reported as `false` when running through a git hook (with `husky`), and `true` when directly running _lint-staged_.

I'm not sure what caused this to start happen, but I assume it is an update to one of:
1. dependencies of _lint-staged_ (colorette, supports-color..., listr2)
2. husky
3. git itself

**Comment by iiroj** [comment]
For a simple reproduction, try running in your terminal:

```
> node -e "console.log(require('tty').isatty(1))"
true
```

Putting this same script into the `pre-commit` sh script of Husky, will print `false` instead:

```sh
#!/usr/bin/env sh
. "$(dirname "$0")/_/husky.sh"

node -e "console.log(require('tty').isatty(1))"
```

**Comment by axmrnv** [comment]
TL;DR: Adding `exec 1> /dev/tty` or `exec >/dev/tty 2>&1` to your git hook before `npx lint-staged` looks like a workaround:

```bash
#!/usr/bin/env sh
. "$(dirname -- "$0")/_/husky.sh"

exec >/dev/tty 2>&1

npx lint-staged
```

---

As @iiroj pointed out, `tty.isatty(1)` returns false if running in the git hook. In order to test it, we have a file `./.husky/pre-commit` like this:

```bash
#!/usr/bin/env sh
. "$(dirname -- "$0")/_/husky.sh"

node -e "console.log(require('tty').isatty(1))"
```

When executing the file directly from bash, it returns `true`; but if it runs in the git hook, it returns `false`:

```console
> bash ./.husky/pre-commit
true

> git hook run pre-commit
false
```

Let's redirect stdout (fd 1) to `/dev/tty`:

```bash
#!/usr/bin/env sh
. "$(dirname -- "$0")/_/husky.sh"

exec 1> /dev/tty

node -e "console.log(require('tty').isatty(1))"
```

Now it returns `true` in both cases:

```console
> bash ./.husky/pre-commit
true

> git hook run pre-commit
true
```

So, adding `exec 1> /dev/tty` (to redirect stdout) or `exec >/dev/tty 2>&1` (to redirect stdout and stderr) to git hook (before payload) looks like a workaround.

**Comment by Ethorsen** [comment]
@axmrnv Amazing discovery.  We'll give it a try, thank you very much for this workaround

**Comment by valentinoli** [comment]
@axmrnv Thank you, it works!

**Comment by iiroj** [comment]
Can somebody open a PR for adding this to the README, maybe under _Frequently Asked Questions_? 🙏

**Comment by Josh-Cena** [comment]
It still appears largely like a bug to me—would appreciate it if anyone could dig out the actual reason why it's suddenly not a tty...

**Comment by iiroj** [comment]
It's not anything done directly in _lint-staged_, as @axmrnv explained pretty well.

I would like to fix it in _lint-staged_, but it would have to be pretty robust. Preferably it's something the `git` team would fix directly so that the default STDOUT would be a TTY.

**Comment by axmrnv** [comment]
Yes, it's git. I don't know why git hook's stdout is not a tty. Maybe there is a reason for that.

In theory, we could fix (workaround) it here by adding `exec >/dev/tty 2>&1` redirection to husky's hook file template. It this case, the redirection would be added to any git hook created by husky. The template is in the followind listing, and redirection might be added beween lines 72 and 74: 

https://github.com/typicode/husky/blob/840c4164d06d3c412caf63a9c9d7ca1af15f2165/src/index.ts#L71-L75

However, this redirection feels a bit hacky, and I'm not sure if it's ok to make it a part of out-of-a-box solution. It's better to tie it to optional CLI parameter (i. e. add redirection if some argument to `husky add` is set). But I don't know whether this solution is aligned to husky owner's vision. @typicode, what do you think?

**Comment by ltbyce** [comment]
@axmrnv Thank you very much!

**Comment by sanzhar-dan** [comment]
Thank you @axmrnv ! It finally fixed that issue.

**Comment by valentinoli** [comment]
My colleague tried to commit in Magit in emacs with the [above workaround](#issuecomment-1176848345) and got the following error:

```
0 git.exe … "add" "-u" "--" "experiences/.husky/pre-commit"
  1 git.exe … "commit" "--"
experiences/.husky/pre-commit: line 5: /dev/tty: No such device or address
husky - pre-commit hook exited with code 1 (error)
```

**Comment by sandumorari** [comment]
> My colleague tried to commit in Magit in emacs with the [above workaround](#issuecomment-1176848345) and got the following error:
> 
> ```
> 0 git.exe … "add" "-u" "--" "experiences/.husky/pre-commit"
>   1 git.exe … "commit" "--"
> experiences/.husky/pre-commit: line 5: /dev/tty: No such device or address
> husky - pre-commit hook exited with code 1 (error)
> ```

Same here

**Comment by valentinoli** [comment]
> My colleague tried to commit in Magit in emacs with the [above workaround](#issuecomment-1176848345) and got the following error

I think we found a workaround for it, but might not work for all environments:

if [ -t 2 ]; then
  exec >/dev/tty 2>&1
fi

based on [this answer](https://stackoverflow.com/a/54535749/8238129)

**Comment by axmrnv** [comment]
@valentinoli, thanks for pointing this out! However, this workaround may be improved. It should work most of the time as `stdout` & `stderr` often go hand in hand; but that's not always the case. If you're going to redirect `stdout` _and_ `stderr`, it's better to test if _both_ files are opened on a terminal:

```bash
if [[ -t 1 && -t 2 ]]; then exec >/dev/tty 2>&1; fi
```

If you want to redirect `stdout` (but not `stderr`), you need to test fd 1 only:

```bash
if [ -t 1 ]; then exec 1> /dev/tty; fi
```

You also may test and redirect `stdout` and `stderr` separately, but I believe it's an overkill:

```bash
if [ -t 1 ]; then exec 1> /dev/tty; fi
if [ -t 2 ]; then exec 2> /dev/tty; fi
```

**Comment by KrishAgarwal2811** [comment]
Would this work for windows powershell / cmd?

**Comment by reywright** [comment]
One note: you may run into commands ran with `lint-staged` that need specific flags added to those commands, to preserve the formatting of them.

For instance, if you run `tsc` commands through `lint-staged`, adding the `--pretty` flag (`tsc --pretty`) may fix your issue. It worked for me, and there may be similar commands you may run to run through `lint-staged` where you have to go expressly throw on a formatting flag that would normally run by default, but not with `lint-staged`.

**Comment by jetersen** [comment]
@axmrnv the conditional approach does not work for windows or well it does but it forces lint-staged into non color mode 😓

```shell
if [[ -t 1 && -t 2 ]]; then exec >/dev/tty 2>&1; fi
```

Also tried but again non color mode.
```
if [ -t 1 ]; then exec 1> /dev/tty; fi
if [ -t 2 ]; then exec 2> /dev/tty; fi
```

Seems to work for windows and uses colors (imo prefer this as it is less verbose). I am not sure why this is the case
```shell
if [ -t 2 ]; then
  exec >/dev/tty 2>&1
fi
```

**Comment by rszalski** [comment]
Just ran into the same thing at work when upgrading `husky`. I noticed that the output of my hooks differs from that of my colleague. 

Thanks to the prior comments hinting at the issue being that the git hooks are not attached to a TTY I wanted to check if `git` version mismatch was the issue. Turns out yes, I was running 2.37.1 and my colleague 2.36.1. 


Git has fixed what I believe to be this exact issue in v 2.37.0: https://github.com/git/git/commit/a082345372e14ecd30a84a790eda0c38154e4602.

Release notes: https://github.com/git/git/blob/master/Documentation/RelNotes/2.37.0.txt#L285

**Comment by akamfoad** [comment]
Hi, I have a similar issue but in my case the terminal is being exploded with repeated log output.

```
OS: Windows
Shell: PowerShell 7.2.6
Git Version: 2.37.1.windows.1
```

running 

```shell
yarn lint-staged
```

#### Through git hooks (added with husky):

https://user-images.githubusercontent.com/41629832/188649709-a61d2579-319a-487b-b2cf-abb89b75cace.mp4

#### Directly running the command:

https://user-images.githubusercontent.com/41629832/188650212-0fa8b4ea-bc04-43f7-9c88-1d68045e96a4.mp4

**Comment by lalitlsg** [comment]
If I make a commit from vscode UI,  by keeping `exec >/dev/tty 2>&1` in pre-commit script
It gives below error:

 .husky/pre-commit: line 4: /dev/tty: Device not configured

**Comment by J4v4Scr1pt** [comment]
> Hi, I have a similar issue but in my case the terminal is being exploded with repeated log output.
> 
> ```
> OS: Windows
> Shell: PowerShell 7.2.6
> Git Version: 2.37.1.windows.1
> ```
> 
> running
> 
> ```shell
> yarn lint-staged
> ```
> 
> #### Through git hooks (added with husky):
>  PowerShell.2022-09-06.16-37-59.mp4 
> #### Directly running the command:
>  PowerShell.2022-09-06.16-39-07.mp4

Exactly the same thing happens to me and my team. Can't understand why and there seems to no solution from what I can find.

**Comment by Nick-Hopps** [comment]
![image](https://user-images.githubusercontent.com/16911282/192674760-3d87e4d6-c8a4-485c-851e-90394809a227.png)

seems they fix it in version 2.38?

**Comment by yangliguo7** [comment]
> 
maybe you should run your command in terminal not in VSCODE or webStrom .

**Comment by akmaldira** [comment]
> Hi, I have a similar issue but in my case the terminal is being exploded with repeated log output.
> 
> ```
> OS: Windows
> Shell: PowerShell 7.2.6
> Git Version: 2.37.1.windows.1
> ```
> 
> running
> 
> ```shell
> yarn lint-staged
> ```
> 
> #### Through git hooks (added with husky):
>  PowerShell.2022-09-06.16-37-59.mp4 
> #### Directly running the command:
>  PowerShell.2022-09-06.16-39-07.mp4

Same problem here

**Comment by MrFlyHouse** [comment]
> > If I make a commit from vscode UI, by keeping `exec >/dev/tty 2>&1` in pre-commit script It gives below error:
> > .husky/pre-commit: line 4: /dev/tty: Device not configured
> 
> @lalitlsg
> 
> i have the same problem but after i change /dev/tty to /dev/console it works
> 
> i dont know how it works exactly but anyway it works. (ㅠ.ㅠ)
> 
> cf: https://unix.stackexchange.com/questions/692257/why-no-such-device-or-address-when-open-dev-tty-in-the-first-process/

@wherehows It works when I'm using terminal panel, but doesn't work when using the SOURCE CONTROL  gui panel (commit by press 'Commit' button), do you have same problem, and do you solve it if so?

VSCode said `.husky/pre-commit: line 4: /dev/console: No such device or address`

**Comment by wherehows** [comment]
> > > If I make a commit from vscode UI, by keeping `exec >/dev/tty 2>&1` in pre-commit script It gives below error:
> > > .husky/pre-commit: line 4: /dev/tty: Device not configured
> > 
> > 
> > @lalitlsg
> > i have the same problem but after i change /dev/tty to /dev/console it works
> > i dont know how it works exactly but anyway it works. (ㅠ.ㅠ)
> > cf: https://unix.stackexchange.com/questions/692257/why-no-such-device-or-address-when-open-dev-tty-in-the-first-process/
> 
> @wherehows It works when I'm using terminal panel, but doesn't work when using the SOURCE CONTROL gui panel (commit by press 'Commit' button), do you have same problem, and do you solve it if so?
> 
> VSCode said `.husky/pre-commit: line 4: /dev/console: No such device or address`

not only source control but also 'fork' tool doesnt work. i haven't solved it yet ㅠ.ㅠ

**Comment by siarheiyelin** [comment]
I managed to solve the issue by downgrading "lint-staged" from 13.x to 12.x

**Comment by iiroj** [comment]
@siarheiyelin Are you sure downgrading _lint-staged_ fixes it? If so, I can take a look at the updates/changes in version 13 and try out if a culprit can be found.

**Comment by iiroj** [comment]
Can you try installing `lint-staged@next` (version 13.1.3) to see if it helps?

**Comment by loriswit** [comment]
> Can you try installing `lint-staged@next` (version [13.1.3](https://github.com/okonet/lint-staged/releases/tag/v13.1.3)) to see if it helps? I restored `supports-color` in place of `colorette`'s `isColorSupported` util, a change that was introduced in 13.0.0.

This fixes the repeated logs issue (as seen in https://github.com/typicode/husky/issues/968#issuecomment-1238185342), but shows instead the uncolored output as described in the original post. I guess this is better than the repeated outputs, but we're basically back to where we started.

**Comment by Geczy** [comment]
![image](https://user-images.githubusercontent.com/1036968/223752926-b92e2751-b2cd-4571-8368-c0c4c34ee8ed.png)

same problem here

> VSCode said .husky/pre-commit: line 4: /dev/console: No such device or address

**Comment by Geczy** [comment]
https://github.com/typicode/husky/issues/968#issuecomment-1455110738

this comment fixed the original issue for me. i reverted my pre-commit file

**Comment by parloti** [comment]
> TL;DR: Adding `exec 1> /dev/tty` or `exec >/dev/tty 2>&1` to your git hook before `npx lint-staged` looks like a workaround:
> 
> ```shell
> #!/usr/bin/env sh
> . "$(dirname -- "$0")/_/husky.sh"
> 
> exec >/dev/tty 2>&1
> 
> npx lint-staged
> ```


This works on windows but if lint-staged fails the commit is not interrupted.

Does anyone have an idea how to get around this?
Would it be possible to capture the lint-stage return and continue or cancel?

**Comment by sbiliaiev** [comment]
@webbertakken Thank you! Your script is exactly what I was looking for! I've tried settings manually `FORCE_COLOR `to `0` and `1` as it's written in other guides but it didn't help me.
Cheers!

**Comment by mreduar** [comment]
> Documented a combination of fixes in https://gist.github.com/webbertakken/c2b457d39224baf701c8de1589b61555.
> 
> Works for:
> 
> * 🍎 MacOS using Zsh (cli) or IDE (no tty)
> * 💻 Ubuntu using Zsh (cli) or IDE (no tty)
> * 🪟 Windows using Git Bash or PowerShell or IDE (no tty)
> 
> Tested with WebStorm IDE and latest git versions.
> 
> @Geczy @akmaldira @MrFlyHouse @loriswit this should solve your problems as well.

Unfortunately it does not work with Windows Terminal WSL under a docker container. For example using Laravel Sail.

**Comment by its-monotype** [comment]
@typicode It's not fixed in v9? I'm facing the same issue...
